### ISSUE TITLE
[feature] Include build/run environment variables in --format=json output

### ISSUE THREAD

**Issue by sfackler** [conan-io/conan#13258]
### What is your suggestion?

The json generator in Conan 1 included the environment variables set by the installed packages in the deps_env_info section. This generator was replaced by the `--format=json` flag in Conan 2, but its output does not include environment variables. They are present in the conanbuildenv/conanrunenv scripts, but those are more painful to properly parse out programmatically than JSON.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @sfackler 

Thanks for your suggestion.
I think adding ``self.dependencies["xxx"].buildenv_info`` and ``self.dependencies["xxx"].runenv_info`` to the serialized node representation of the graph that goes in the ``--format=json`` (I guess you meant ``conan install/create/graph --format=json`` output, did you?)

There are a couple of challenges to serialize that, because they represent multiple operations, like append, prepend, unset, etc, so it is not as easy as serializing a list of values, if we want to do it correctly, but lets give it a try.

**Comment by sfackler** [comment]
Yep, I'm interested in `conan install`'s output.

While it'd definitely be good to be able to split this out granularly per-dependency with each operation specified, my personal use-case would be satisfied with a top-level "resolved" form of the environment variables like what ends up in the coanbuildenv/conanrunenv scripts. I think that was how the old `deps_env_info` output behaved as well.

It seems like something like this could work to represent the fully precise output:

```json
"runenv_info": {
    "FOOBAR": [
        { "type": "append", "value": "fizzbuzz" }
    ]
}
```

**Comment by memsharded** [comment]
I think the aggregated representation would be difficult to represent in the graph json. The ``cpp_info`` is already serialized, but not aggregated, because the aggregation actually only happens now at ``VirtualBuildEnv`` and ``VirtualRunEnv`` generators, and this is even sometimes customized via ``generate()`` calls. 

Aggregating things at the graph level with ``deps_cpp_info`` and ``deps_env_info`` was quite limiting and problematic, and it was deprecated in 1.X in favor of ``self.dependencies``, and now finally removed in 2.0. So there is no longer any generic aggregation of upstream values in the graph. Furthermore, this would be quite costly to evaluate for every node in the graph, so this is not something that can be done easily. Computing only for the root node would be quite asymmetric. 

I think dumping environment consumer side could be more a responsibility of a ``deployer``, but it depends what you want to do. What usage are you trying to do with that information?

**Comment by sfackler** [comment]
I'm using Conan to manage native dependencies in a broader build system compiling Rust code, so I want to be able to run `conan install`, and then run e.g. `cargo build` or `cargo test` with the environment configured based off of the Conan buildenv. I essentially want to be able to do `source conan_install_dir/conanbuild.sh && cargo test` but with some more programmatic control and not having to add an extra layer of bash in the command invoation.

A deployer may make more sense for this than the `--forma=json` output.

**Comment by petermbauer** [comment]
I ran into the same issue. My use-case is to provide the build and/or run environment in terms of Jenkins' [withEnv](https://www.jenkins.io/doc/pipeline/steps/workflow-basic-steps/#withenv-set-environment-variables). With Conan 1.x, we used the JSON info about the environment variables to build up the parameter for `withEnv`.
I would also be willing to implement a generator but this does not seem to be a trivial thing to implement in a generic way.

**Comment by memsharded** [comment]
Thanks for the feedback @petermbauer 

I'd like to understand better the use case. Conan already generates environment scripts for end-consumer conanfiles (when the process is driven by ``conan build`` or ``conan create``), so instead of:

```
vars = read(conangenerated.vars)
withEnv({dict from vars}:
     sh mycommand
```
it could be:
```
sh source conanbuild.sh && mycommand
````

**Comment by petermbauer** [comment]
Sure, but this requires that the command being executed can be manipulated.
Indirectly executing a command via a function like this is not possible:
```
def someFunction() {
    sh "mycommand"
}

withEnv() {
    someFunction()
}
```
While "someFunction()" could also be defined in some library or plugin.

Thanks for your time!

**Comment by memsharded** [comment]
> While "someFunction()" could also be defined in some library or plugin.

Is that your library or plugin? Isn't it possible to do something like:
```
def someFunction(env_file=None) {
    if env_file:
         sh source env_file && "mycommand"
    else:
         sh "mycommand"
}
```
Or is it some kind of closed source library?

In any case, creating a file with key=value  for env-vars should be straightforward, check https://docs.conan.io/2/reference/tools/env/envvars.html#iterating-the-variables, something like:

```python
def generate(self):
     envvars = VirtualBuildEnv(self).vars()
     save(self, "myenvfile", "\n".join([f"{k}={v}" for k, v in envvars.items()]
```
Haven't you tried something like that?

**Comment by petermbauer** [comment]
Yes, i am aware of these alternatives, but the the point is about having a generic way to execute any tool/library/plugin in the build/run environment defined by a "conan install" without having to rely on some special code in the respective conanfile.
This is what was possible with the information provided in the Conan 1.x JSON output.

Its not a blocker for sure, and if it does not fit into the big picture anymore or it is not possible to implement in a clean way, i will adapt and work around it, but the underlying assumption was that if the env var information can be "encoded" into shell scripts already, it would be relatively easy to put it out in a "machine-readable" format like JSON.

**Comment by memsharded** [comment]
> Its not a blocker for sure, and if it does not fit into the big picture anymore or it is not possible to implement in a clean way, i will adapt and work around it, but the underlying assumption was that if the env var information can be "encoded" into shell scripts already, it would be relatively easy to put it out in a "machine-readable" format like JSON.

Sure, I am not saying this is not possible, and in fact this is something that we have been wanting to investigate to add it to Conan for some time.
But it is very important for us to understand the limitations of the current available solutions, where do they fail short, what exact formats are needed (because you ask for JSON, but then other users might want other different formats, and then the feature starts to be a bit too cluttered), are we sure it is not better to have other format, because Jenkins can already get that directly from a file with a specific format? Adding things in the output graph.json might be possible, but then other users will complain that they have to load and parse the json, and that is not ergonomic, etc.
So I am not discussing to not add it, but quite the opposite, I am interested in adding something, but this feature has not been added so far because the details were not clear enough, so I am trying to learn and understand better the use cases.

**Comment by petermbauer** [comment]
I see, thanks for the clarification!
JSON is easy to handle in Jenkins with `readJSON()`. Getting it from the output instead of a file is nice as it does not pollute the build/install/deploy folder so maybe a format specifier like `--format json-env` or `json-with-env` would make sense to keep it separate.

**Comment by memsharded** [comment]
I have been investigating a bit more. It seems that if we are dumping env-vars to a file, the best would be to comply with the almost standard dotenv ``.env`` files: https://github.com/motdotla/dotenv, there are tools for different languages to manage it, there are plugins for IDEs, etc. 

It looks like this wouldn't belong to the graph json output (we could dump there the ``buildenv_info`` values provided by packages, but not the resulting aggregation for the consumer, that belongs mostly to generators). So I think this could belong to the ``VirtualBuildEnv`` and ``VirtualRunEnv`` generators, likely enabling it with a conf like ``tools.env:use_env_files`` or the like.

**Comment by petermbauer** [comment]
Sounds good as well. One thing i was wondering is if/how calling the `vcvarsall.bat` could be implemented or substituted. I don't have a need or use case for it so far though.

**Comment by memsharded** [comment]
Hi all,

Conan added the ``tools.env:dotenv`` to create dotenv "standard" files besides the specific OS env-var activation scripts.
Would this be the solution to your needs @sfackler @petermbauer ?

**Comment by petermbauer** [comment]
@memsharded yes, we are already using it for the debugging with VS Code, thanks a lot!

**Comment by memsharded** [comment]
Closing this ticket as staled, and most likely solved by ``tools.env:dotenv``.
Please dont hesitate to comment or create new tickets for any further questions or issues. Thanks for the feedback!
